Tracker alignment package (a work in progress)

ryuwd opened the pull request:
My work on Tracker alignment (with no-field cosmics) up to now. This package is a work in progress - the code is unpolished and I'm still debugging and developing diagnostics to converge towards a working alignment fit. I don't expect any of my changes here to change nightly validation.

Apologies for this huge PR. After this, I plan to submit smaller PR's more regularly.

I will have made some mistakes or broken convention so if you spot anything please feel free to leave reviews and comments.  I have summarised the contents of this PR and how everything is supposed to fit together below.

### Summary of changes
- Merged latest master as of 17dc9d690b74418a5f8aad99309f1591a8904934. Resolved some conflicts (none major)

- CosmicReco:
   - trailing whitespace removed in some files during merge conflict resolution
   - fix header guard in `MinuitDriftFitter`
   - PDFFit.cc: residual calculations now use signed DOCA 
   - `LineFinder` now uses the aligned tracker
- RecoDataProducts:
   - (CosmicTrackSeed.cc) changed a forward declaration to declare TimeCluster (a struct) as a struct, not a class


- JobConfig:
   - Two job fcl's added in `alignment/`. `select_nofield_cosmictracks.fcl` was used to generate DS-cosmic-nofield-alignselect in MDC2020Dev.

- TrackerAlignment:
   - New package
   - new art filter module AlignTrackSelector: used to generate `DS-cosmic-nofield-alignselect`
   - new art analyzer module AlignTrackCollector: contains detailed diagnostics and outputs Millepede alignment fit data to file using tracks from CosmicReco time fits **WIP**
   - new script: aligntrack_display.py: takes ROOT files generated by AlignTrackCollector jobs and outputs track diagnostics (see flowchart LHS) **WIP**
   - new script: generate_derivatives.py: symbolic algebra utility (uses SymPy) used to generate `AlignmentDerivatives.cc` and `AlignmentDerivatives.hh` . - **needs to be made more presentable (WIP)**
   - new script : generate_derivatives_test.py: unit test skeleton for extra validation - diagnostics in AlignTrackCollector are quite thorough at the moment but plan to complete this soon **WIP**
   - Test folder: contains misalignment test files and test job fcl's

### Expected alignment workflow
![Alignment flow(2)](https://user-images.githubusercontent.com/56410978/82936768-fa2e6500-9f86-11ea-81fe-b9f0bf20e842.png)
- 'Digis' refers to art files containing 'digi' data products. I am using DS-cosmic-nofield or DS-cosmic-nofield-alignselect
- Track Reco uses Richie's Time Fit ( [Mu2e-doc-33162](https://mu2e-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=33162) )
- `Mille` is a helper class provided by the Millepede alignment package to write input files for the `PEDE` executable, which performs the alignment fit.
- 'Align Tracker' refers to the stage carried out by `AlignedTrackerMaker` (TrackerConditions). It uses alignment constants provided to the Proditions interface to change tracker straw positions accordingly.
- 'Track Collection' is where the `AlignTrackCollector` module selects tracks, calculates the needed quantities for alignment, and writes the data to file using `Mille`.
- 'PEDE' refers to the millepede executable that performs the alignment fit given the output file(s) produced by the `Mille` class during the `AlignTrackCollector` job(s).

### Issues (not a complete list)
- `AlignTrackCollector_module.cc`: has gotten very long - mostly due to messy and repetitive diagnostics. I'd like to clean up and move most of the diagnostics to its own module soon once the alignment fit is converging properly.
- `AlignTrackCollector_module.cc`: uses ROOT matrix classes derived from TObject - performance issues? not sure about this
- AlignedTrackerMaker (TrackerConditions): panel center of rotation as `Panel::straw0Midpoint()` may not be ideal. Panel DOFs are disabled for now but I plan to look into this eventually
- (generally) cleanup required
- Residual covariance calculation sometimes yields 'covariance' matrices that are not positive semidefinite.

### To-do before merge:
...

FNALbuild commented:
Hi @ryuwd,
You have proposed changes to files in these packages:
- RecoDataProducts
- JobConfig
- TrackerAlignment
- CosmicReco

which require these tests: build.




<a href="https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild">FNALbuild is explained here.</a>

ryuwd commented:
Thanks Sophie!

I pushed a few more commits to fix some regressions / bugs I spotted after merging in latest master. (also, some code cleanup)

bonventre commented on CosmicReco/src/PDFFit.cc:
```diff
@@ -336,12 +336,66 @@ double GaussianDriftFit::DOCAresidual(ComboHit const& sh, const std::vector<doub
-  return predictedDistance-measuredDistance;
+  double resid = predictedDistance - measuredDistance;
+
+  return (pca.s1() > 0 ? resid : -resid);
```
I'm not sure I follow this line, is it supposed to be signing by the ambiguity?

ryuwd replied:
Yes. I meant to ask you about this further but didn't get around to it:

This is based on `TwoLinePCA_XYZ::LRambig()`, which uses `s2() > 0 ? 1 : -1` but I use `s1()` instead of `s2()` here because TwoLinePCA in the time fit is constructed like this: `TwoLinePCA(straw_p, straw_d, track_p, track_d)` while earlier use of TwoLinePCA_XYZ in CosmicReco is constructed like this `TwoLinePCA_XYZ(track_p, track_d, straw_p, straw_d)`.

I sign the residual because if I only sign the predictedDistance I can get large time residuals ~80ns and lots of tracks end up getting rejected.

Maybe @brownd1978 can confirm if I'm doing this correctly? Maybe `TwoLinePCA` should also have an `LRambig()` function.

bonventre replied:
Right, I remember now - I wasn't sure if this was correct, I thought s1 was the distance along the straw direction, i.e. whether the track was on the Cal or HV side of the straw

ryuwd replied:
Ah... this might need some checking then. In this snippet 
https://github.com/Mu2e/Offline/blob/bd2e7884502edfa02712645a29f482fb11431a2f/CosmicReco/src/DriftFitUtils.cc#L119-L138

The track information is passed before the straw information. s1() and s2() are different if you swap the order around like this
https://github.com/Mu2e/Offline/blob/bd2e7884502edfa02712645a29f482fb11431a2f/CosmicReco/src/PDFFit.cc#L302-L304
...so I'm not sure which is right. Maybe the best thing to do is just to be consistent?

bonventre replied:
I dont think s2 is right either, since thats just which side of the track startposition you are on - I think you would need to do something like https://github.com/Mu2e/Offline/blob/master/TrkDiag/src/KalDiag.cc#L481

ryuwd replied:
I followed Dave's suggestion from earlier.  Happy to try this method as well

ryuwd replied:
I wrote a HitAmbiguity function implementing the suggested method:
https://github.com/Mu2e/Offline/blob/d0db6280233b105ec5e8f2b9f746926253edf01f/CosmicReco/src/PDFFit.cc#L375-L393

Does this make sense? I've not tested it yet.

bonventre commented on TrackerAlignment/inc/Mille.h:
```diff
@@ -0,0 +1,78 @@
+ *  License along with this program (see the file COPYING.LIB for more
+ *  details); if not, write to the Free Software Foundation, Inc.,
+ *  675 Mass Ave, Cambridge, MA 02139, USA.
+ */
```
Someone should check that this is all compatible with Offline's current apache license

ryuwd replied:
Ah, I'm not sure about this either. Someone else will have to answer. 

If necessary I can write my own class to write the Millepede files. It should be fairly straightforward to do and `Mille` as it is now doesn't support compression or writing doubles. (but the 'pede' software does support reading compressed files of doubles)

bonventre commented on TrackerAlignment/src/AlignmentDerivatives.cc:
```diff
@@ -0,0 +1,2189 @@
+CosmicTrack_DCA_Deriv_panel_g(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel)};
+return result;
+}
+
```
Should this be in gen? (especially if it might change)

ryuwd replied:
I believe `gen/` is reserved for code which is generated during the build, so I'm not sure I'd like to re-generate this code every time the code is rebuilt. I think it is safest if generated once, validated, then kept as it is. 

It should be regenerated if `generate_derivatives.py` changes, usually to reflect if something in `AlignedTrackerMaker` changes. It is also slow to run (it takes around 30 seconds to a minute)

brownd1978 commented:
Is there a reason the TwoLinePoca is constructed opposite the convention
that normally signs DOCA?
I believe in Richies ML fit the sign of DOCA doesn't matter, but he should
confirm.  In the Kalman fit and any
fit relying on analytic derivatives DOCA must be signed to avoid
discontinuous 1st derivatives.

On Wed, May 27, 2020 at 4:51 PM ryuwd <notifications@github.com> wrote:

> *@ryuwd* commented on this pull request.
> ------------------------------
>
> In CosmicReco/src/PDFFit.cc
> <https://github.com/Mu2e/Offline/pull/201#discussion_r431506342>:
>
> >    double measuredDistance = srep.driftTimeToDistance(sh.strawId(), sh.time()-hit_t0, 0);
>
> -  return predictedDistance-measuredDistance;
> +  double resid = predictedDistance - measuredDistance;
> +
> +  return (pca.s1() > 0 ? resid : -resid);
>
> Yes. I meant to ask you about this further but didn't get around to it:
>
> This is based on TwoLinePCA_XYZ::LRambig(), which uses s2() > 0 ? 1 : -1
> but I use s1() instead of s2() here because TwoLinePCA in the time fit is
> constructed like this: TwoLinePCA(straw_p, straw_d, track_p, track_d)
> while earlier use of TwoLinePCA_XYZ in CosmicReco is constructed like this TwoLinePCA_XYZ(track_p,
> track_d, straw_p, straw_d).
>
> I sign the residual because if I only sign the predictedDistance I can get
> large time residuals ~80ns and lots of tracks end up getting rejected.
>
> Maybe @brownd1978 <https://github.com/brownd1978> can confirm if I'm
> doing this correctly? Maybe TwoLinePCA should also have an LRambig()
> function.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/201#discussion_r431506342>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH573H5EG6MUSVAU2IO53RTWRPLANCNFSM4NLAIBCA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

ryuwd commented:
Hi Dave,

Thanks, I was unsure about this up to now - I believe you're right and I should restore the TwoLinePCA construction to be consistent with the rest of the codebase.

Best wishes, Ryun

kutschke commented:
Hi Everyone,

    Are any of these functions more generic than GaussianDriftFit?  Will we end up duplicating them in other places?  If so, we should refactor them out of this class into a new class that can be used in other places.  If you go this way you should also consider free functions rather than member functions, and making Straw const& a function argument.


    Rob


________________________________
From: ryuwd <notifications@github.com>
Sent: Thursday, May 28, 2020 11:35 AM
To: Mu2e/Offline <Offline@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [Mu2e/Offline] Tracker alignment package (a work in progress) (#201)


@ryuwd commented on this pull request.

________________________________

In CosmicReco/src/PDFFit.cc<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_201-23discussion-5Fr431972217&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=BN0tpx1nIYHAGSbLGTG6OEwY89CdK19T88aThofNpOE&s=mm5zSGH1eoqDA7kso6IBzsIFVtOtDVf2UB7USa8iKwg&e=>:

>    double measuredDistance = srep.driftTimeToDistance(sh.strawId(), sh.time()-hit_t0, 0);

-  return predictedDistance-measuredDistance;
+  double resid = predictedDistance - measuredDistance;
+
+  return (pca.s1() > 0 ? resid : -resid);


I wrote a HitAmbiguity function implementing the suggested method:
https://github.com/Mu2e/Offline/blob/d0db6280233b105ec5e8f2b9f746926253edf01f/CosmicReco/src/PDFFit.cc#L375-L393<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_blob_d0db6280233b105ec5e8f2b9f746926253edf01f_CosmicReco_src_PDFFit.cc-23L375-2DL393&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=BN0tpx1nIYHAGSbLGTG6OEwY89CdK19T88aThofNpOE&s=Iy3jwuht34TovCX-rjMFp-Oc_jSQWAiF8BGjfRijgYA&e=>

Does this make sense? I've not tested it yet.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_201-23discussion-5Fr431972217&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=BN0tpx1nIYHAGSbLGTG6OEwY89CdK19T88aThofNpOE&s=mm5zSGH1eoqDA7kso6IBzsIFVtOtDVf2UB7USa8iKwg&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABHY5ZUJKKEQ453CVHH3FPDRT2HFRANCNFSM4NLAIBCA&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=Zi5IH-k2_JiK_Wipiv3pwwSXO9UNaHiLP8sisfPz__k&m=BN0tpx1nIYHAGSbLGTG6OEwY89CdK19T88aThofNpOE&s=KJ1w7Y1Tm9S3tA89z_rttouV1JJxQH4-iWzWPZWlQeY&e=>.

brownd1978 commented:
For a single hit + track fit iteration you can estimate LRambig with this
formula, but for the helix
fit we factored this out so it could be assigned externally with more
powerful algorithms.  I don't know if you want/need to do that for the
cosmic fit.

On Thu, May 28, 2020 at 9:41 AM ryuwd <notifications@github.com> wrote:

> *@ryuwd* commented on this pull request.
> ------------------------------
>
> In CosmicReco/src/PDFFit.cc
> <https://github.com/Mu2e/Offline/pull/201#discussion_r431972217>:
>
> >    double measuredDistance = srep.driftTimeToDistance(sh.strawId(), sh.time()-hit_t0, 0);
>
> -  return predictedDistance-measuredDistance;
> +  double resid = predictedDistance - measuredDistance;
> +
> +  return (pca.s1() > 0 ? resid : -resid);
>
> I wrote a HitAmbiguity function implementing the suggested method:
>
> https://github.com/Mu2e/Offline/blob/d0db6280233b105ec5e8f2b9f746926253edf01f/CosmicReco/src/PDFFit.cc#L375-L393
>
> Does this make sense? I've not tested it yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/201#discussion_r431972217>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH576WISEXEGJ6IEVA7M3RT2HFPANCNFSM4NLAIBCA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

sophiemiddleton commented:
Sorry, trying to catch up with the comments here. I think we never looked at an external Ambiguity Resolver for Cosmics. I think we decided a simple thing (like we see here) would be sufficient, but as this becomes more sophisticated we need to make sure that's consistent with what we expect in KinKal. Ryun does that AMBIG come from the utility I made - I think it does -?

brownd1978 commented:
KinKal supports pluggable ambiguity resolution algorithm, just like BTrk,
so once the cosmic fit is working there we will get this functionality for
free.  I'm not sure it's worth developing a parallel technology for the ML
fit.

On Thu, May 28, 2020 at 10:40 AM Sophie Middleton <notifications@github.com>
wrote:

> Sorry, trying to catch up with the comments here. I think we never looked
> at an external Ambiguity Resolver for Cosmics. I think we decided a simple
> thing (like we see here) would be sufficient, but as this becomes more
> sophisticated we need to make sure that's consistent with what we expect in
> KinKal. Ryun does that AMBIG come from the utility I made - I think it does
> -?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/201#issuecomment-635495005>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572BCCSLBDTPGQ2UPW3RT2OZRANCNFSM4NLAIBCA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

bonventre commented:
Since for the ML fit this is a diagnostic only, you actually can only use the simple ambiguity formulation - anything else will give you residuals that will not be consistent with the fit

sophiemiddleton commented:
True, I think I only made that function to make those plots to show the True_Ambig/Reco_Ambig. It was for diagnostics only.

ryuwd commented:
>  Ryun does that AMBIG come from the utility I made - I think it does -?

Yes. It's based on LRambig() from TwoLinePCA_XYZ (which is used in DriftFitUtils)

ryuwd commented:
I added a bunch of new commits just now. The changes involve some major cleanup in configuration and track collector module. It also includes a working version of my MilleDataWriter class, which supports double precision and gzip compression of Millepede input files. ( I have tested it ) All source code from millepede is now removed

ryuwd commented:
Latest fixes/features:
- Millepede results are translated correctly to AlignedTracker Proditions table text files
- Alignment job helper (scripts/setup.sh) supports running track collection jobs in parallel

ryuwd commented:
Apologies for the many commits added since opening this PR. 

Based on some more recent tests running Millepede and 'Track collection', I think this PR is in a good state to be reviewed and merged from this point onwards.

brownd1978 commented:
Hi Ryun,
did you write this code, or was it machine written?  If machine written,
please include a reference and/or explain how it was generated.
I see lines for 6 planes and 6 panels, Will this code need to extended when
we add more DOFs to the alignment?
thanks, Dave

On Mon, Jun 1, 2020 at 6:14 AM ryuwd <notifications@github.com> wrote:

> *@ryuwd* commented on this pull request.
> ------------------------------
>
> In TrackerAlignment/src/AlignmentDerivatives.cc
> <https://github.com/Mu2e/Offline/pull/201#discussion_r433225633>:
>
> > +{
> +        std::vector<double> result = {CosmicTrack_DCA_Deriv_plane_dx(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_plane_dy(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_plane_dz(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_plane_a(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_plane_b(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_plane_g(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_dx(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_dy(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_dz(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_a(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_b(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel),
> +CosmicTrack_DCA_Deriv_panel_g(a0,b0,a1,b1,t0,plane_dx,plane_dy,plane_dz,plane_a,plane_b,plane_g,panel_dx,panel_dy,panel_dz,panel_a,panel_b,panel_g,wire_x,wire_y,wire_z,wdir_x,wdir_y,wdir_z,plane_x,plane_y,plane_z,panel_straw0x,panel_straw0y,panel_straw0z,driftvel)};
> +return result;
> +}
> +
>
> I believe gen/ is reserved for code which is generated during the build,
> so I'm not sure I'd like to re-generate this code every time the code is
> rebuilt. I think it is safest if generated once, validated, then kept as it
> is.
>
> It should be regenerated if generate_derivatives.py changes, usually to
> reflect if something in AlignedTrackerMaker changes. It is also slow to
> run (it takes around 30 seconds to a minute)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/201#discussion_r433225633>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57ZOBM4BY4GHZSX52VTRUOSRPANCNFSM4NLAIBCA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

ryuwd commented:
Hi Dave,

With these functions, hit residual derivatives can be calculated given any straw in any plane and panel. You just need to pass the relevant information about the geometry (the current alignment constants, plane and panel origins, etc)
I can include some explanations in my README file if that's adequate

ryuwd commented:
ryuwd/Offline#44 included some refactoring and reformatting of CosmicReco to make sure const& references are used throughout. Also, some reformatting as well.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/201/commits/9d52b49c992bc97750e347ca69fe6f544beae2f6: build

FNALbuild commented:
:sunny: The tests passed at ref 9d52b49c992bc97750e347ca69fe6f544beae2f6.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/scons.log). Build time: 12 min 27 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (5) FIXME (8) in 13 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 14 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//136/console).

bonventre commented on CosmicReco/src/PDFFit.cc:
```diff
@@ -1,359 +1,443 @@
+    double doca = calculate_DOCA(this->chits[i], a0, a1, b0, b1, this->tracker);
+    double time_residual = this->TimeResidual(doca, this->srep, t0, this->chits[i], this->tracker);
+    double pdf_t = 1 / sqrt(2 * TMath::Pi() * this->sigma_t * this->sigma_t) *
+                   exp(-(time_residual * time_residual) / (2 * this->sigma_t * this->sigma_t));
```
I'm going to suggest that going forward we have any large scale reformatting put in separate pull requests - it just makes a review of this much code harder

ryuwd commented:
ryuwd/Offline#46 resolves hit signs using `HitAmbiguity` in `PDFFit.cc`.
Also:
- cleaned up SConscript
- some intermediate progress on `AlignTrackDiagnostics` - will significantly shorten AlignTrackCollector when finished

ryuwd commented:
Analytical derivatives now use HitAmbiguity as well.

brownd1978 commented:
@FNALbuild run build test

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/201/commits/bf172f2f9261e06f013a0999993d6ccf20bd2f2e: build

FNALbuild commented:
:sunny: The tests passed at ref bf172f2f9261e06f013a0999993d6ccf20bd2f2e.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/scons.log). Build time: 11 min 58 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :wavy_dash: | [TODO (5) FIXME (9) in 16 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 14 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//138/console).
